### PR TITLE
Flag dismiss logging

### PR DESCRIPTION
**Pull request by imorland**
Sorry @franzliedke - this is something I left out of the the PR by mistake 😞 

Rather than deleting flags from the DB after they have been dealt with, we have a requirement to be able to see an audit trail of flags raised by users, who actioned them, and when.

We've been running this code as a git patch for a while now on our production site.

Currently there is no UI to see the audit trail, but we do reference the data from various reports. Longer term goal is to create a 'closed flag' tab, which will use the data collected by this PR.

### CODE REVIEW

File: src/Command/DeleteFlagsHandler.php
Comment (franzliedke) [inline src/Command/DeleteFlagsHandler.php]: Could we do this with a single query, rather than looping through the objects in PHP and creating a query for each of them?

**Review by franzliedke — CHANGES_REQUESTED** [review]
I'd say we need to discuss first whether we want this feature by default.

The "principle of data economy" may be relevant here? cc @jordanjay29

File: src/Command/DeleteFlagsHandler.php
Comment (imorland) [reply src/Command/DeleteFlagsHandler.php]: Yes, absolutely. Bad on my part for not making that more efficient.

**Comment by imorland** [comment]
If data economy is an issue here, I'd be happy to wrap this behind a preferences toggle?

**Comment by franzliedke** [comment]
Yeah, we need to discuss it and give it some thought. I'll predict this won't quite make the cut for beta.12 (we still have too many other things that were *planned* for that release), but we'll be sure to look at it for the following one.

**Comment by luceos** [comment]
I think this is a feature we cannot go without. Economy of data shouldn't be an issue yet and one can always create a command that wipes data beyond a specific age.

**Comment by imorland** [comment]
What's the thoughts here now? If I commit the requested changes, are we good to proceed with this one?

**Comment by franzliedke** [comment]
> Economy of data shouldn't be an issue yet 

I was referring to a GDPR principle but apparently misnamed it - the idea of only storing the data you need and nothing more. I was not referring to space or scaling issues (the "yet" makes me think that's how you interpreted it). 

Of course, if internal processes require you to keep and process this data, that would be a legitimate need.

@jordanjay29 Do we have this use case as well? Do we want a setting to make this optional?

**Comment by franzliedke** [comment]
Personally, I guess it makes sense, and would vote for an option.

@luceos Would you feel this is in line with the idea of not adding opinionated features (as outlined in our currently-WIP core extension policy)? I'd say if there's an option, it is configurable and thus not as opinionated. (Of course we don't want to go over board with options, either.)

I will review the code (it already looks good, just would prefer some stuff moved elsewhere or encapsulated).

**Comment by luceos** [comment]
This being an extension (and not core) gives us the freedom to shape its direction more easily. People can fork and modify if they disagree.

Having said so I do think this should be a toggle/option and the feature should have the suggested tab to look at the trail. In my opinion we should hold off on merging this without these two. I know this sounds unfair for a feature that makes a lot of sense, but a feature that isn't entirely usable by a layman isn't something we should publish as 1st party team. 

PS Franz, I love how you asked this exact question, because it made me more aware of the implications involved with merging such a feature 🤗

**Comment by jordanjay29** [comment]
> Do we have this use case as well? Do we want a setting to make this optional?

Definitely, @franzliedke. This would help a lot when posts get approved on accident, or to track past misbehavior of repeat rules offenders.

**Comment by franzliedke** [comment]
> I know this sounds unfair for a feature that makes a lot of sense, but a feature that isn't entirely usable by a layman isn't something we should publish as 1st party team.

To be honest, I think it's a reasonable approach for an end-user product such as Flarum. I didn't even think about this, because I often push for small, iterative improvements (even when they're not user-visible) in my day job.

@imorland I hope you can understand this line of reasoning. Can we ask you for a minimal UI to make this available? It does not need to be perfect (*that* we can iterate on), but if we add this code, it should at least result in user-visible data.

(In the long term, I could see this integrated into a "Moderation" page / UI provided by core.)

**Comment by imorland** [comment]
@franzliedke absolutely understand. I'll look into it very soon. Are you thinking some UI within `/admin` for this, or forum side?

**Comment by franzliedke** [comment]
For the short-term solution, a page in the admin UI should be fine, IMO. Not sure if a full-blown admin section is warranted, but as we don't yet have established patterns for sub-pages in the admin area, I'll leave that open. :wink:

**Comment by luceos** [comment]
That's quite some work on that last commit @imorland ; well done, thanks. This now needs someone to clone and confirm all to be in order for us to approve a merge.

**Comment by imorland** [comment]
Credit to @andrewdyer and @harrygiffgaff for that last commit 🥇

**Comment by luceos** [comment]
Merging checklist (let's see if this works)


- [x] Priority: (~~critical, current cycle, future cycle, backlog~~, unplanned)
_Do we need this? If it is unplanned, we might need to close, as this is often best delegated to 3rd party extensions or future versions._
- [x] The PR was made to the appropriate repository. 
_Some changes should be part of a 1st party extension instead of core and vice versa._
- [ ] All tests are green.
- [ ] Core developer confirmed locally this works as intended.
_Does a check out locally provide the desired effect? Does nothing else seem broken?_
- [x] Adheres to our conventions or can be patched up easily after merging.
- [x] Does not touch any lines outside of the intended changes, eg through formatting or compilation.
- [ ] Has the necessary companion PR's, eg for translations.

Ref: https://github.com/flarum/org/issues/47#issuecomment-600829092

**Comment by imorland** [comment]
How are we looking here now @luceos @franzliedke ?

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
Looks great. thanks a lot! One small thing: can we use the Model extender for relationships that have been added? This should also resolve the merge conflict.

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
Overall, works great (checked out locally)! A few UI suggestions though:

- Can we link usernames to the user's profile page?
- Can we add a bit of padding to the left of the table? Not that critical an issue, but looks a bit cramped.
- Can we allow the table to scroll on overflow? (both in x and y directions)
- By default, I think flags should be sorted from most recently dismissed to last recently dismissed. Although if it would be possivle to add sorting options to the table's header row, that would be really nice!
- We need the accompanying PR to https://github.com/flarum/lang-english/pulls
- StyleCI

**Comment by imorland** [comment]
lang-english is here https://github.com/flarum/lang-english/pull/160

- [X] Can we link usernames to the user's profile page?
- [X] Can we add a bit of padding to the left of the table? Not that critical an issue, but looks a bit cramped.
- [X] Can we allow the table to scroll on overflow? (both in x and y directions)
- [X] By default, I think flags should be sorted from most recently dismissed to last recently dismissed. Although if it would be possivle to add sorting options to the table's header row, that would be really nice!
- [X] We need the accompanying PR to https://github.com/flarum/lang-english/pulls
- [X] StyleCI

image

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
I have a few more concerns, sorry :speak_no_evil:

- The table still doesn't scroll in the x-direction. I realize that having a scroll bar in the middle of the screen (for tables that don't take up the whole page) isn't the prettiest (as a side note, and this is unrelated to this PR, but core team should probably look into making the admin page always full height), but currently, you can't see the whole table on smaller-screened devices.

- Can we add a tooltip to the date columns to see the actual datetime when the flag was created / dismissed?

File: js/src/admin/components/DismissedFlagsTable.js
Comment (dsevillamartin) [inline js/src/admin/components/DismissedFlagsTable.js:35]: We could perhaps turn this into
```js
['username', 'discussion', 'created_at', ...]
  .map(k => <th>...</th>)
```

Maybe even extensible - add a method that returns the keys in an ItemList or regular array and loop them for columns ?

Could perhaps do the same with the tbody below.

Not sure if we want this at all, just throwing ideas.

Comment (andrewdyer) [reply js/src/admin/components/DismissedFlagsTable.js:35]: I would go against the mapping idea. In my opinion, it doesn't make sense to have the columns in the header row dynamically generated from an array when the columns in the table body rows are hard defined. I think we'd end up creating more overhead/confusion for ourselves.

One thing to note is the final column in the header row is empty, therefore unless we decide that the final column should have a label we'd end up wiring some logic to skip over rendering any text output. I feel this suggestion is finding a solution to an issue which is not there. Also, this page on the admin panel was supposed to be a [short term solution](https://github.com/flarum/flags/pull/21#issuecomment-596014300).

Comment (dsevillamartin) [reply js/src/admin/components/DismissedFlagsTable.js:35]: Well that's why I said perhaps make the tbody extensible as well. If it's a temporary solution tho, it's probs fine then.

Also - you could have an empty <th> after the mapped array (that wouldn't exclude it).

**Review by askvortsov1 — APPROVED** [review]
Needs some empty lines at ends of files, but otherwise good to go imo!

**Comment by imorland** [comment]
Discovered an issue with this today, where dismissed flags re-present themselves after a page refresh. We/I will get that addressed ASAP

**Review by askvortsov1 — APPROVED** [review]

**Review by tankerkiller125 — APPROVED** [review]

**Comment by askvortsov1** [comment]
At some point in 1.x, we'll be generalizing the user table from https://github.com/flarum/core/pull/2626 into a generic model table. Let's wait for that change before proceeding here.

**Comment by askvortsov1** [comment]
Hello, thank you for your pull request! In order to speed up future development, keep all work in one place, and take advantage of CI tools that can help us avoid breaking changes, we have moved all Flarum code to a single monorepo at [flarum/framework](https://github.com/flarum/framework). You can read more about this process [in our dev diary](https://discuss.flarum.org/d/29894-dev-diary-build-20/3). Unfortunately, pull requests can't be carried over to the monorepo, so we have to close all open pull requests. If this pull request is still relevant, please feel free to reopen it over on the monorepo. We apologize for the inconvenience, and hope you will consider contributing to Flarum again in the future.

**Pull request closed without merging**
